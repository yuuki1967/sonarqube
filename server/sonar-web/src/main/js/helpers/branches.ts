/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sortBy } from 'lodash';

export function isBranch(branchLike?: T.BranchLike): branchLike is T.Branch {
  return branchLike !== undefined && (branchLike as T.Branch).isMain !== undefined;
}

export function isShortLivingBranch(branchLike?: T.BranchLike): branchLike is T.ShortLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as T.ShortLivingBranch).type === 'SHORT'
  );
}

export function isLongLivingBranch(branchLike?: T.BranchLike): branchLike is T.LongLivingBranch {
  return (
    isBranch(branchLike) && !branchLike.isMain && (branchLike as T.LongLivingBranch).type === 'LONG'
  );
}

export function isMainBranch(branchLike?: T.BranchLike): branchLike is T.MainBranch {
  return isBranch(branchLike) && branchLike.isMain;
}

export function isPullRequest(branchLike?: T.BranchLike): branchLike is T.PullRequest {
  return branchLike !== undefined && (branchLike as T.PullRequest).key !== undefined;
}

export function getPullRequestDisplayName(pullRequest: T.PullRequest) {
  return `${pullRequest.key} â€“ ${pullRequest.title}`;
}

export function getBranchLikeDisplayName(branchLike: T.BranchLike) {
  return isPullRequest(branchLike) ? getPullRequestDisplayName(branchLike) : branchLike.name;
}

export function getBranchLikeKey(branchLike: T.BranchLike) {
  return isPullRequest(branchLike) ? `pull-request-${branchLike.key}` : `branch-${branchLike.name}`;
}

export function getBranchQualityGateColor(status: string) {
  let indicatorColor = 'gray';
  if (status === 'ERROR') {
    indicatorColor = 'red';
  } else if (status === 'WARN') {
    indicatorColor = 'orange';
  } else if (status === 'OK') {
    indicatorColor = 'green';
  }
  return indicatorColor;
}

export function isSameBranchLike(a: T.BranchLike | undefined, b: T.BranchLike | undefined) {
  // main branches are always equal
  if (isMainBranch(a) && isMainBranch(b)) {
    return true;
  }

  // short- and long-living branches are compared by type and name
  if (
    (isLongLivingBranch(a) && isLongLivingBranch(b)) ||
    (isShortLivingBranch(a) && isShortLivingBranch(b))
  ) {
    return a.type === b.type && a.name === b.name;
  }

  // pull requests are compared by id
  if (isPullRequest(a) && isPullRequest(b)) {
    return a.key === b.key;
  }

  // finally if both parameters are `undefined`, consider them equal
  return a === b;
}

export function sortBranchesAsTree(branchLikes: T.BranchLike[]) {
  const result: T.BranchLike[] = [];

  const mainBranch = branchLikes.find(isMainBranch);
  const longLivingBranches = branchLikes.filter(isLongLivingBranch);
  const shortLivingBranches = branchLikes.filter(isShortLivingBranch);
  const pullRequests = branchLikes.filter(isPullRequest);

  // main branch is always first
  if (mainBranch) {
    result.push(
      mainBranch,
      ...getPullRequests(mainBranch.name),
      ...getNestedShortLivingBranches(mainBranch.name)
    );
  }

  // then all long-living branches
  sortBy(longLivingBranches, 'name').forEach(longLivingBranch => {
    result.push(
      longLivingBranch,
      ...getPullRequests(longLivingBranch.name),
      ...getNestedShortLivingBranches(longLivingBranch.name)
    );
  });

  // finally all orhpan pull requests and branches
  result.push(
    ...sortBy(pullRequests.filter(pr => pr.isOrphan), pullRequest => pullRequest.key),
    ...sortBy(shortLivingBranches.filter(branch => branch.isOrphan), branch => branch.name)
  );

  return result;

  /** Get all short-living branches (possibly nested) which should be merged to a given branch */
  function getNestedShortLivingBranches(mergeBranch: string) {
    const found: T.ShortLivingBranch[] = shortLivingBranches.filter(
      branch => branch.mergeBranch === mergeBranch
    );

    let i = 0;
    while (i < found.length) {
      const current = found[i];
      found.push(...shortLivingBranches.filter(branch => branch.mergeBranch === current.name));
      i++;
    }

    return sortBy(found, branch => branch.name);
  }

  function getPullRequests(base: string) {
    return sortBy(pullRequests.filter(pr => pr.base === base), pullRequest => pullRequest.key);
  }
}

export function getBranchLikeQuery(branchLike?: T.BranchLike): T.BranchParameters {
  if (isShortLivingBranch(branchLike) || isLongLivingBranch(branchLike)) {
    return { branch: branchLike.name };
  } else if (isPullRequest(branchLike)) {
    return { pullRequest: branchLike.key };
  } else {
    return {};
  }
}

// Create branch object from branch name or pull request key
export function fillBranchLike(
  branch?: string,
  pullRequest?: string
): T.ShortLivingBranch | T.PullRequest | undefined {
  if (branch) {
    return {
      isMain: false,
      mergeBranch: '',
      name: branch,
      type: 'SHORT'
    } as T.ShortLivingBranch;
  } else if (pullRequest) {
    return { base: '', branch: '', key: pullRequest, title: '' } as T.PullRequest;
  }
  return undefined;
}
