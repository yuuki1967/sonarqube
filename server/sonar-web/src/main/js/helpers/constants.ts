/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as theme from '../app/theme';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT'
];
export const RULE_TYPES: T.RuleType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT',
  'UNKNOWN'
];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  theme.green,
  theme.lightGreen,
  theme.yellow,
  theme.orange,
  theme.red
];
export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  theme.red,
  theme.orange,
  theme.yellow,
  theme.lightGreen,
  theme.green
];

export const RATING_COLORS = [theme.green, theme.lightGreen, theme.yellow, theme.orange, '#e00'];
