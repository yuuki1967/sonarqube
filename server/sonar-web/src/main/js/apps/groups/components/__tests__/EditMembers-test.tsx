/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { mount } from 'enzyme';
import EditMembers from '../EditMembers';
import { click, waitAndUpdate } from '../../../../helpers/testUtils';

it('should edit members', async () => {
  const group = { id: 3, name: 'Foo', membersCount: 5 };
  const onEdit = jest.fn();

  const wrapper = mount(<EditMembers group={group} onEdit={onEdit} organization="org" />);
  expect(wrapper).toMatchSnapshot();

  click(wrapper.find('ButtonIcon'));
  expect(wrapper).toMatchSnapshot();

  await waitAndUpdate(wrapper);
  click(wrapper.find('ResetButtonLink'));
  expect(onEdit).toBeCalled();
  expect(wrapper).toMatchSnapshot();
});
