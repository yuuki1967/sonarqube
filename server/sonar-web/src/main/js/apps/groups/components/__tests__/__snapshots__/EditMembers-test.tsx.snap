// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should edit members 1`] = `
<EditMembers
  group={
    Object {
      "id": 3,
      "membersCount": 5,
      "name": "Foo",
    }
  }
  onEdit={[MockFunction]}
  organization="org"
>
  <ButtonIcon
    className="button-small"
    onClick={[Function]}
  >
    <Button
      className="button-small button-icon"
      onClick={[Function]}
      stopPropagation={true}
      style={
        Object {
          "color": "#236a97",
        }
      }
    >
      <button
        className="button button-small button-icon"
        onClick={[Function]}
        style={
          Object {
            "color": "#236a97",
          }
        }
        type="button"
      >
        <BulletListIcon>
          <Icon>
            <svg
              height={16}
              style={
                Object {
                  "clipRule": "evenodd",
                  "fillRule": "evenodd",
                  "strokeLinejoin": "round",
                  "strokeMiterlimit": 1.41421,
                }
              }
              version="1.1"
              viewBox="0 0 16 16"
              width={16}
              xmlSpace="preserve"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            >
              <path
                d="M2.968 11.274v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM2.968 8.255v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM2.968 5.235v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM15.045 11.274v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM2.968 2.216v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM15.045 8.255v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM15.045 5.235v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM15.045 2.216v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177z"
                style={
                  Object {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
          </Icon>
        </BulletListIcon>
      </button>
    </Button>
  </ButtonIcon>
</EditMembers>
`;

exports[`should edit members 2`] = `
<EditMembers
  group={
    Object {
      "id": 3,
      "membersCount": 5,
      "name": "Foo",
    }
  }
  onEdit={[MockFunction]}
  organization="org"
>
  <ButtonIcon
    className="button-small"
    onClick={[Function]}
  >
    <Button
      className="button-small button-icon"
      onClick={[Function]}
      stopPropagation={true}
      style={
        Object {
          "color": "#236a97",
        }
      }
    >
      <button
        className="button button-small button-icon"
        onClick={[Function]}
        style={
          Object {
            "color": "#236a97",
          }
        }
        type="button"
      >
        <BulletListIcon>
          <Icon>
            <svg
              height={16}
              style={
                Object {
                  "clipRule": "evenodd",
                  "fillRule": "evenodd",
                  "strokeLinejoin": "round",
                  "strokeMiterlimit": 1.41421,
                }
              }
              version="1.1"
              viewBox="0 0 16 16"
              width={16}
              xmlSpace="preserve"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            >
              <path
                d="M2.968 11.274v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM2.968 8.255v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM2.968 5.235v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM15.045 11.274v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM2.968 2.216v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM15.045 8.255v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM15.045 5.235v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM15.045 2.216v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177z"
                style={
                  Object {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
          </Icon>
        </BulletListIcon>
      </button>
    </Button>
  </ButtonIcon>
  <EditMembers
    group={
      Object {
        "id": 3,
        "membersCount": 5,
        "name": "Foo",
      }
    }
    onClose={[Function]}
    organization="org"
  >
    <Modal
      contentLabel="users.update"
      onRequestClose={[Function]}
    >
      <Modal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        className="modal"
        closeTimeoutMS={0}
        contentLabel="users.update"
        isOpen={true}
        onRequestClose={[Function]}
        overlayClassName="modal-overlay"
        parentSelector={[Function]}
        portalClassName="ReactModalPortal"
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={true}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
      >
        <Portal
          containerInfo={
            <div
              class="ReactModalPortal"
            >
              <div
                class="ReactModal__Overlay ReactModal__Overlay--after-open modal-overlay"
              >
                <div
                  aria-label="users.update"
                  class="ReactModal__Content ReactModal__Content--after-open modal"
                  role="dialog"
                  tabindex="-1"
                >
                  <header
                    class="modal-head"
                  >
                    <h2>
                      users.update
                    </h2>
                  </header>
                  <div
                    class="modal-body"
                  >
                    <div
                      class="select-list"
                    >
                      <div
                        class="display-flex-center"
                      >
                        <ul
                          class="radio-toggle spacer-right"
                        >
                          <li>
                            <input
                              checked=""
                              id="filter__selected"
                              name="filter"
                              type="radio"
                              value="selected"
                            />
                            <label
                              for="filter__selected"
                            >
                              selected
                            </label>
                          </li>
                          <li>
                            <input
                              id="filter__deselected"
                              name="filter"
                              type="radio"
                              value="deselected"
                            />
                            <label
                              for="filter__deselected"
                            >
                              unselected
                            </label>
                          </li>
                          <li>
                            <input
                              id="filter__all"
                              name="filter"
                              type="radio"
                              value="all"
                            />
                            <label
                              for="filter__all"
                            >
                              all
                            </label>
                          </li>
                        </ul>
                        <div
                          class="search-box"
                          title=""
                        >
                          <input
                            aria-label="search_verb"
                            autocomplete="off"
                            class="search-box-input"
                            maxlength="100"
                            placeholder="search_verb"
                            type="search"
                            value=""
                          />
                          <svg
                            class="search-box-magnifier"
                            height="16"
                            style="fill-rule: evenodd; stroke-linejoin: round; stroke-miterlimit: 1.41421;"
                            version="1.1"
                            viewBox="0 0 16 16"
                            width="16"
                            xml:space="preserve"
                            xmlns:xlink="http://www.w3.org/1999/xlink"
                          >
                            <path
                              d="M10.308 7.077c0-.89-.316-1.65-.949-2.283a3.111 3.111 0 0 0-2.282-.948c-.89 0-1.65.316-2.283.948a3.111 3.111 0 0 0-.948 2.283c0 .89.316 1.65.948 2.282a3.111 3.111 0 0 0 2.283.949c.89 0 1.65-.316 2.282-.949a3.111 3.111 0 0 0 .949-2.282zm3.692 6c0 .25-.091.466-.274.649a.887.887 0 0 1-.65.274.857.857 0 0 1-.648-.274L9.954 11.26c-.86.596-1.82.894-2.877.894a4.989 4.989 0 0 1-1.972-.4 5.076 5.076 0 0 1-1.623-1.082A5.076 5.076 0 0 1 2.4 9.049 4.989 4.989 0 0 1 2 7.077c0-.688.133-1.345.4-1.972a5.076 5.076 0 0 1 1.082-1.623A5.076 5.076 0 0 1 5.105 2.4 4.989 4.989 0 0 1 7.077 2c.687 0 1.345.133 1.972.4a5.076 5.076 0 0 1 1.623 1.082c.454.454.815.995 1.082 1.623.266.627.4 1.284.4 1.972a4.938 4.938 0 0 1-.894 2.877l2.473 2.474a.883.883 0 0 1 .267.649z"
                              style="fill: currentColor;"
                            />
                          </svg>
                          
                        </div>
                      </div>
                      <div
                        class="select-list-list-container spacer-top"
                      >
                        <ul
                          class="menu"
                        />
                      </div>
                    </div>
                  </div>
                  <footer
                    class="modal-foot"
                  >
                    <button
                      class="button button-link"
                      type="reset"
                    >
                      Done
                    </button>
                  </footer>
                </div>
              </div>
            </div>
          }
        >
          <ModalPortal
            ariaHideApp={true}
            bodyOpenClassName="ReactModal__Body--open"
            className="modal"
            closeTimeoutMS={0}
            contentLabel="users.update"
            defaultStyles={
              Object {
                "content": Object {
                  "WebkitOverflowScrolling": "touch",
                  "background": "#fff",
                  "border": "1px solid #ccc",
                  "borderRadius": "4px",
                  "bottom": "40px",
                  "left": "40px",
                  "outline": "none",
                  "overflow": "auto",
                  "padding": "20px",
                  "position": "absolute",
                  "right": "40px",
                  "top": "40px",
                },
                "overlay": Object {
                  "backgroundColor": "rgba(255, 255, 255, 0.75)",
                  "bottom": 0,
                  "left": 0,
                  "position": "fixed",
                  "right": 0,
                  "top": 0,
                },
              }
            }
            isOpen={true}
            onRequestClose={[Function]}
            overlayClassName="modal-overlay"
            parentSelector={[Function]}
            portalClassName="ReactModalPortal"
            role="dialog"
            shouldCloseOnEsc={true}
            shouldCloseOnOverlayClick={true}
            shouldFocusAfterRender={true}
            shouldReturnFocusAfterClose={true}
            style={
              Object {
                "content": Object {},
                "overlay": Object {},
              }
            }
          >
            <div
              className="ReactModal__Overlay ReactModal__Overlay--after-open modal-overlay"
              onClick={[Function]}
              onMouseDown={[Function]}
              style={Object {}}
            >
              <div
                aria-label="users.update"
                className="ReactModal__Content ReactModal__Content--after-open modal"
                onClick={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                role="dialog"
                style={Object {}}
                tabIndex="-1"
              >
                <header
                  className="modal-head"
                >
                  <h2>
                    users.update
                  </h2>
                </header>
                <div
                  className="modal-body"
                >
                  <DeferredSpinner
                    loading={true}
                    timeout={100}
                  >
                    <SelectList
                      elements={Array []}
                      onSearch={[Function]}
                      onSelect={[Function]}
                      onUnselect={[Function]}
                      renderElement={[Function]}
                      selectedElements={Array []}
                    >
                      <div
                        className="select-list"
                      >
                        <div
                          className="display-flex-center"
                        >
                          <RadioToggle
                            className="spacer-right"
                            disabled={false}
                            name="filter"
                            onCheck={[Function]}
                            options={
                              Array [
                                Object {
                                  "disabled": false,
                                  "label": "selected",
                                  "value": "selected",
                                },
                                Object {
                                  "disabled": false,
                                  "label": "unselected",
                                  "value": "deselected",
                                },
                                Object {
                                  "disabled": false,
                                  "label": "all",
                                  "value": "all",
                                },
                              ]
                            }
                            value="selected"
                          >
                            <ul
                              className="radio-toggle spacer-right"
                            >
                              <li
                                key="selected"
                              >
                                <input
                                  checked={true}
                                  disabled={false}
                                  id="filter__selected"
                                  name="filter"
                                  onChange={[Function]}
                                  type="radio"
                                  value="selected"
                                />
                                <Tooltip>
                                  <label
                                    htmlFor="filter__selected"
                                  >
                                    selected
                                  </label>
                                </Tooltip>
                              </li>
                              <li
                                key="deselected"
                              >
                                <input
                                  checked={false}
                                  disabled={false}
                                  id="filter__deselected"
                                  name="filter"
                                  onChange={[Function]}
                                  type="radio"
                                  value="deselected"
                                />
                                <Tooltip>
                                  <label
                                    htmlFor="filter__deselected"
                                  >
                                    unselected
                                  </label>
                                </Tooltip>
                              </li>
                              <li
                                key="all"
                              >
                                <input
                                  checked={false}
                                  disabled={false}
                                  id="filter__all"
                                  name="filter"
                                  onChange={[Function]}
                                  type="radio"
                                  value="all"
                                />
                                <Tooltip>
                                  <label
                                    htmlFor="filter__all"
                                  >
                                    all
                                  </label>
                                </Tooltip>
                              </li>
                            </ul>
                          </RadioToggle>
                          <SearchBox
                            autoFocus={true}
                            loading={false}
                            onChange={[Function]}
                            placeholder="search_verb"
                            value=""
                          >
                            <div
                              className="search-box"
                              title=""
                            >
                              <input
                                aria-label="search_verb"
                                autoComplete="off"
                                autoFocus={true}
                                className="search-box-input"
                                maxLength={100}
                                onChange={[Function]}
                                onKeyDown={[Function]}
                                placeholder="search_verb"
                                type="search"
                                value=""
                              />
                              <DeferredSpinner
                                loading={false}
                                timeout={100}
                              >
                                <SearchIcon
                                  className="search-box-magnifier"
                                >
                                  <Icon
                                    className="search-box-magnifier"
                                  >
                                    <svg
                                      className="search-box-magnifier"
                                      height={16}
                                      style={
                                        Object {
                                          "clipRule": "evenodd",
                                          "fillRule": "evenodd",
                                          "strokeLinejoin": "round",
                                          "strokeMiterlimit": 1.41421,
                                        }
                                      }
                                      version="1.1"
                                      viewBox="0 0 16 16"
                                      width={16}
                                      xmlSpace="preserve"
                                      xmlnsXlink="http://www.w3.org/1999/xlink"
                                    >
                                      <path
                                        d="M10.308 7.077c0-.89-.316-1.65-.949-2.283a3.111 3.111 0 0 0-2.282-.948c-.89 0-1.65.316-2.283.948a3.111 3.111 0 0 0-.948 2.283c0 .89.316 1.65.948 2.282a3.111 3.111 0 0 0 2.283.949c.89 0 1.65-.316 2.282-.949a3.111 3.111 0 0 0 .949-2.282zm3.692 6c0 .25-.091.466-.274.649a.887.887 0 0 1-.65.274.857.857 0 0 1-.648-.274L9.954 11.26c-.86.596-1.82.894-2.877.894a4.989 4.989 0 0 1-1.972-.4 5.076 5.076 0 0 1-1.623-1.082A5.076 5.076 0 0 1 2.4 9.049 4.989 4.989 0 0 1 2 7.077c0-.688.133-1.345.4-1.972a5.076 5.076 0 0 1 1.082-1.623A5.076 5.076 0 0 1 5.105 2.4 4.989 4.989 0 0 1 7.077 2c.687 0 1.345.133 1.972.4a5.076 5.076 0 0 1 1.623 1.082c.454.454.815.995 1.082 1.623.266.627.4 1.284.4 1.972a4.938 4.938 0 0 1-.894 2.877l2.473 2.474a.883.883 0 0 1 .267.649z"
                                        style={
                                          Object {
                                            "fill": "currentColor",
                                          }
                                        }
                                      />
                                    </svg>
                                  </Icon>
                                </SearchIcon>
                              </DeferredSpinner>
                            </div>
                          </SearchBox>
                        </div>
                        <SelectListListContainer
                          disabledElements={Array []}
                          elements={Array []}
                          filter="selected"
                          onSelect={[Function]}
                          onUnselect={[Function]}
                          renderElement={[Function]}
                          selectedElements={Array []}
                        >
                          <div
                            className="select-list-list-container spacer-top"
                          >
                            <ul
                              className="menu"
                            />
                          </div>
                        </SelectListListContainer>
                      </div>
                    </SelectList>
                  </DeferredSpinner>
                </div>
                <footer
                  className="modal-foot"
                >
                  <ResetButtonLink
                    onClick={[Function]}
                  >
                    <ButtonLink
                      onClick={[Function]}
                      type="reset"
                    >
                      <Button
                        className="button-link"
                        onClick={[Function]}
                        type="reset"
                      >
                        <button
                          className="button button-link"
                          onClick={[Function]}
                          type="reset"
                        >
                          Done
                        </button>
                      </Button>
                    </ButtonLink>
                  </ResetButtonLink>
                </footer>
              </div>
            </div>
          </ModalPortal>
        </Portal>
      </Modal>
    </Modal>
  </EditMembers>
</EditMembers>
`;

exports[`should edit members 3`] = `
<EditMembers
  group={
    Object {
      "id": 3,
      "membersCount": 5,
      "name": "Foo",
    }
  }
  onEdit={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  organization="org"
>
  <ButtonIcon
    className="button-small"
    onClick={[Function]}
  >
    <Button
      className="button-small button-icon"
      onClick={[Function]}
      stopPropagation={true}
      style={
        Object {
          "color": "#236a97",
        }
      }
    >
      <button
        className="button button-small button-icon"
        onClick={[Function]}
        style={
          Object {
            "color": "#236a97",
          }
        }
        type="button"
      >
        <BulletListIcon>
          <Icon>
            <svg
              height={16}
              style={
                Object {
                  "clipRule": "evenodd",
                  "fillRule": "evenodd",
                  "strokeLinejoin": "round",
                  "strokeMiterlimit": 1.41421,
                }
              }
              version="1.1"
              viewBox="0 0 16 16"
              width={16}
              xmlSpace="preserve"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            >
              <path
                d="M2.968 11.274v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM2.968 8.255v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM2.968 5.235v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM15.045 11.274v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM2.968 2.216v1.51q0 0.102-0.075 0.177t-0.177 0.075h-1.51q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h1.51q0.102 0 0.177 0.075t0.075 0.177zM15.045 8.255v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM15.045 5.235v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177zM15.045 2.216v1.51q0 0.102-0.075 0.177t-0.177 0.075h-10.568q-0.102 0-0.177-0.075t-0.075-0.177v-1.51q0-0.102 0.075-0.177t0.177-0.075h10.568q0.102 0 0.177 0.075t0.075 0.177z"
                style={
                  Object {
                    "fill": "currentColor",
                  }
                }
              />
            </svg>
          </Icon>
        </BulletListIcon>
      </button>
    </Button>
  </ButtonIcon>
</EditMembers>
`;
